dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1051 @@
+
+	req := &network.BlockRequestMessage{}
+	testPeerA := peer.ID("a")
+	testPeerB := peer.ID("b")
+	peersTried := make(map[peer.ID]struct{})
+
+	// test base case
+	cs.peerState[testPeerA] = &peerState{
+		number: 129,
+	}
+	cs.peerState[testPeerB] = &peerState{
+		number: 257,
+	}
+
+	peers := cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 2, len(peers))
+	require.Contains(t, peers, testPeerA)
+	require.Contains(t, peers, testPeerB)
+
+	// test peer ignored case
+	cs.ignorePeers[testPeerA] = struct{}{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+
+	// test all peers ignored case
+	cs.ignorePeers[testPeerB] = struct{}{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 2, len(peers))
+	require.Contains(t, peers, testPeerA)
+	require.Contains(t, peers, testPeerB)
+	require.Equal(t, 0, len(cs.ignorePeers))
+
+	// test peer's best block below number case, shouldn't include that peer
+	start, err := variadic.NewUint32OrHash(130)
+	require.NoError(t, err)
+	req.StartingBlock = *start
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+
+	// test peer tried case, should ignore peer already tried
+	peersTried[testPeerA] = struct{}{}
+	req.StartingBlock = variadic.Uint32OrHash{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+}
+
+func Test_chainSync_logSyncSpeed(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	mockBlockState := NewMockBlockState(ctrl)
+	mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+	mockBlockState.EXPECT().GetHighestFinalisedHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+
+	mockNetwork := NewMockNetwork(ctrl)
+	mockNetwork.EXPECT().Peers().Return([]common.PeerInfo{}).AnyTimes()
+
+	type fields struct {
+		blockState  BlockState
+		network     Network
+		state       chainSyncState
+		benchmarker *syncBenchmarker
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "state bootstrap",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       bootstrap,
+			},
+		},
+		{
+			name: "case tip",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       tip,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			cs := &chainSync{
+				ctx:         ctx,
+				cancel:      cancel,
+				blockState:  tt.fields.blockState,
+				network:     tt.fields.network,
+				state:       tt.fields.state,
+				benchmarker: tt.fields.benchmarker,
+			}
+			go cs.logSyncSpeed()
+			time.Sleep(time.Second * 6)
```

Comment:
make the duration of the ticker configurable on `chainSync`, and you'll be able to shorten this test by setting a duration in nanoseconds.

Reply:
yes, that shortens it a lot, updated.

Reply:
nit why `time.Sleep(time.Nanosecond * 2)` and not `time.Sleep(time.Nanosecond)`?

Reply:
I used time.Nanosecond * 2 since log frequency was set for time.Nanosecond, I thought I should give it more time to run, but yes time.Nanosecond works, so I've updated.